### PR TITLE
Using multithreading make loadDefaultTables faster

### LINKED ISSUE #3979: make incubator-shardingsphere first initialization run faster
## Feature Request

**For English only**, other languages will not accept.

Please pay attention on issues you submitted, because we maybe need more details. 
If no response **more than 7 days** and we cannot make decision by current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Is your feature request related to a problem?
#3681 
### Describe the feature you would like.
Using multithreading make load tablemeta faster and eliminate some duplicate loads.
to optimize the code location
image

### PR DESCRIPTION
**Pull request by beijing-penguin**
Fixes #3979 #3681

Changes proposed in this pull request:
- Using multithreading make loadDefaultTables faster

### CODE REVIEW

File: sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/sharding/execute/metadata/loader/ShardingTableMetaDataLoader.java [hunk 1]
```diff
@@ -206,19 +207,46 @@ public TableMetas loadAll(final ShardingRule shardingRule) throws SQLException {
...
+    private Map<String, TableMetaData> loadDefaultTables(final ShardingRule shardingRule, final Map<String, TableMetaData> result) throws SQLException {
         Optional<String> actualDefaultDataSourceName = shardingRule.findActualDefaultDataSourceName();
         if (!actualDefaultDataSourceName.isPresent()) {
             return Collections.emptyMap();
         }
         Collection<String> tableNames = loadAllTableNames(actualDefaultDataSourceName.get());
-        Map<String, TableMetaData> result = new HashMap<>(tableNames.size(), 1);
-        for (String each : tableNames) {
-            result.put(each, load(each, shardingRule));
+        removeRepeatTable(tableNames, result);
```
Comment (tristaZero) [inline sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/sharding/execute/metadata/loader/ShardingTableMetaDataLoader.java]: If there are two tables with the same name `t1` and different structures in ds0 and default ds seperately, do you think removeRepeatTable() could return correct result list?

Comment (beijing-penguin) [reply sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/sharding/execute/metadata/loader/ShardingTableMetaDataLoader.java]: > If there are two tables with the same name `t1` and different structures in ds0 and default ds seperately, do you think removeRepeatTable() could return correct result list?

yes, load() methed  will be get out the same as tableMetaData by table name...loadShardingTables or loadDefaultTables execute load() method and only first parameter table name is different.
image

Comment (beijing-penguin) [reply sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/sharding/execute/metadata/loader/ShardingTableMetaDataLoader.java]: > Why did you remove repeat table? For performance or anything else?
 For performance.

File: sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/sharding/execute/metadata/loader/ShardingTableMetaDataLoader.java [hunk 2]
```diff
@@ -206,19 +207,46 @@ public TableMetas loadAll(final ShardingRule shardingRule) throws SQLException {
...
+            @Override
+            public Collection<TableMetaData> execute(final Collection<String> inputs, final boolean isTrunkThread, final Map<String, Object> dataMap) throws SQLException {
+                String logicTableName = inputs.iterator().next();
+                Collection<TableMetaData> result = new LinkedList<>();
+                result.add(load(logicTableName, shardingRule));
+                return result;
+            }
+        });
+        Object[] tableNameArr = tableNames.toArray();
+        for (int i = 0, size = tableNames.size(); i < size; i++) {
```
[the remark this answers is not in the document]
Comment (beijing-penguin) [reply sharding-core/sharding-core-execute/src/main/java/org/apache/shardingsphere/sharding/execute/metadata/loader/ShardingTableMetaDataLoader.java]: > Why not `for (int i = 0; i < tableNames.size(); i++) `?

Avoid size() method multiple computations,,,for performance

**Pull request merged**